### PR TITLE
[python] [build system] Improve MacOS Wheel Building with Rosetta

### PR DESCRIPTION
**Pull request by zbentley**
**Update**: `pyenv` has been removed from the build chain. While the isolation and tooling it provides is convenient, it proved far too opinionated about reaching out and pulling in Homebrew-installed dependencies to be something that adds to build reproducibility (for example, `gettext` is depended on if it is installed via `brew`, regardless of architecture, causing linker errors and segfaults when Python localizes strings).

Fixes https://github.com/apache/pulsar/issues/15402

### Motivation

Building Python client wheels on MacOS is troublesome on some combinations of architecture/python version. Specifically, Python 3.7 is the most widely used version of Python 3, but it does not support building "universal" binary wheels (wheels which contain both ARM and Intel-architecture binary artifacts).

Additionally, building Python from source on MacOS is troublesome and prone to breakage in older (or future) MacOS versions.

As a result, the wheel building Bash automation refrains from creating 3.7 wheels on ARM, and otherwise special-cases 3.7 in a few different ways.

Two tools are available to help with these issues which the Pulsar client build system does not currently make use of:
- ~The [pyenv](https://github.com/pyenv/pyenv) project is a widely used means of getting around "python has trouble building from source" on many systems, including MacOS systems. They maintain a curated list of patches (mostly to Python's build system rather than Python itself) to allow it to build well on all sorts of environments. They also provide tooling for semi-hermetic "virtual python environments" (distinct from [virtualenvs](https://docs.python.org/3/library/venv.html) in that they encapsulate a separate Python _interpreter_, rather than just a set of dependencies). Using those environments, it's possible to build and install multiple versions of Python and switch between them.~
- The [rosetta](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment) translation system on ARM Macs allows those Macs to run binaries (including compiler toolchains) for both the ARM and x86 architecture.

### Modifications

This PR alters the wheel creation script in a few ways:
- The MacOS SDK (the stdlib and system headers/libraries) is now looked up using the standard `xcrun` command rather than via hardcoded and templated path.
- Python versions iterated over in the script are also qualified with their architecture--x86, arm, or universal.
- The OpenSSL version that is built for the Pulsar client to link against is also used by Python builds.
- ~`pyenv` is installed (via git master checkout) in the wheel building scratch environment. `git clone` is Pyenv's suggested installation mechanism, since many of their patches land in `master` before they are available in pyenv package artifacts via e.g. Homebrew or dpkg repositories.~
- ~`pyenv` is used to install Pythons, rather than a direct download-and-compile.  Appropriate flags are set/unset for each python/architecture "tuple". For example, for 3.7, only native-architecture, non-universal wheels can be created, so its build/install is run twice, one with an ARM architecture and once with x86. **Note that this means that arm64 python37 builds will fail on Intel Macs**; the Rosetta translation layer used to allow ARM macs to run the Intel build system does not reverse, and so the `arch` command's invocation will fail for that permutation on an Intel mac. I suggest that Intel mac compiling users comment that permutation out of the `PYTHON_VERSIONS` list.~
- `pip`, `setuptools`, and `wheel` packages are updated to their latest versions supported on the corresponding Python.
- ~Pyenv is also used to invoke `bdist_wheel` at the end, providing added assurance that wheel creation executes against the desired versions of resources.~

### Verifying this change

I suggest the following test plan:

1. On a Mac running MacOS 11 or better, ensure `cmake` and the SDK are installed. Also install `pyenv` in a separate distribution for testing via `brew install pyenv`.
1. If your test system is an Intel Mac, comment out the python37/arm64 entry from the PYTHON_VERSIONS list in this PR's file.
1. Run `bash build-python-wheels.sh`
1. Ensure that the expected wheel artifacts corresponding to the listed Python versions are created in the `pulsar-client-cpp/python/dist` directory.
1. For each python interpreter version, do `python -mpip install --upgrade pip; python -mpip install --upgrade setuptools wheel`.
1. For each python, find the corresponding wheel file in `dist/` and do `python -mpip install /path/to/pulsar-client-wheel.whl`. Ensure it installs successfully.
1. Ensure that only system-avaialble or relative linker paths are available via `otool -L $(python -c 'import _pulsar; print(_pulsar.__file__)')`.
1. Run the python Pulsar client test suite against a running instance of Pulsar, and ensure all tests pass.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): sort of
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: client libraries

### Documentation

Need to update docs? 

- [x] `no-need-doc` 
The wheel creation process is unchanged externally.

### CODE REVIEW

File: pulsar-client-cpp/python/build-mac-wheels.sh [hunk 1]
```diff
@@ -41,77 +42,36 @@ CURL_VERSION=7.61.0
...
     curl -O -L https://github.com/openssl/openssl/archive/OpenSSL_${OPENSSL_VERSION}.tar.gz
-    # -arch arm64 -arch x86_64
     tar xvfz OpenSSL_${OPENSSL_VERSION}.tar.gz
     mv openssl-OpenSSL_${OPENSSL_VERSION} openssl-OpenSSL_${OPENSSL_VERSION}-arm64
     pushd openssl-OpenSSL_${OPENSSL_VERSION}-arm64
       CFLAGS="-fPIC -mmacosx-version-min=${MACOSX_DEPLOYMENT_TARGET}" \
-          ./Configure --prefix=$PREFIX no-shared darwin64-arm64-cc
+          ./Configure --openssldir=$PREFIX --prefix=$PREFIX no-shared darwin64-arm64-cc
       make -j8
       make install
```
Comment (merlimat) [inline pulsar-client-cpp/python/build-mac-wheels.sh:76]: This could be `install_sw` too

**Review by merlimat — APPROVED** [review]

**Comment by merlimat** [comment]
@zbentley I'm getting errors in building Python 3.7 with Pyenv through the script: 

```
BUILD FAILED (OS X 12.3.1 using python-build 2.3.0-1-g867f34b5)

Inspect or clean up the working tree at /var/folders/q6/sh01_cr95z7f13gwk1thn3y00000gn/T/python-build.20220505154848.25738
Results logged to /var/folders/q6/sh01_cr95z7f13gwk1thn3y00000gn/T/python-build.20220505154848.25738.log

Last 10 log lines:
		case upgrade in \
			upgrade) ensurepip="--upgrade" ;; \
			install|*) ensurepip="" ;; \
		esac; \
		DYLD_LIBRARY_PATH=/var/folders/q6/sh01_cr95z7f13gwk1thn3y00000gn/T/python-build.20220505154848.25738/Python-3.7.13 ./python.exe -E -m ensurepip \
			$ensurepip --root=/ ; \
	fi
dyld[40040]: missing symbol called
/bin/sh: line 1: 40040 Abort trap: 6           DYLD_LIBRARY_PATH=/var/folders/q6/sh01_cr95z7f13gwk1thn3y00000gn/T/python-build.20220505154848.25738/Python-3.7.13 ./python.exe -E -m ensurepip $ensurepip --root=/
make: *** [install] Error 134
```

**Comment by zbentley** [comment]
@merlimat when convenient could you attach one of the pyenv build log files mentioned in the error snippet you posted?

**Comment by zbentley** [comment]
Interesting; that's a new one on me (and on google). As with most things pyenv, it's (over)sensitive to workstation environments and so novel failure modes often crop up when used on machines that are configured to build other software.

 A couple things to try:
* The presence of `python.exe` is a little weird. Some parts of the python-build system use that, but it shows up a lot in your output. Are you by chance running a case-insensitive MacOS filesystem (this is somewhat rare these days). What version of MacOS are you running, what `arch`, and what FS (`diskutil info / | grep -i personality`)?
* Make sure you don't have anything er ... unusual being inherited from your system environment. Pyenv's not terribly hermetic when run by default (are you running it via the build-wheels script on that branch, or normally on your shell?).
* Try removing `--with-openssl=$PREFIX` from this line and building again: https://github.com/apache/pulsar/pull/15436/files#diff-5644e2b9669a87858353608a740e7980b50b7161ce9cd84a3b800ec1c53a0999R205
* Try building with `PYENV_DEBUG=1` and post resulting logs/error output.
* Add `-k` to your `pyenv install` stanza so it keeps the build dir around. That'll allow you to re-run the failing commands with various changes/debugging rather than rebuilding Python each time. Once you've done that, the tmpdir pyenv uses should not be deleted once it's done. You can `export TMPDIR=/stable/path` to make it predictable. Then, try these:
* Try with faulthandler; if the partially-built python is something you can reproduce that issue by hand with (e.g. you can re-run that ensurepip command after the build fails and it still reproduces the dyld error), add `-X faulthandler`, so like `....python.exe -X faulthandler -E -m ensurepip`.
* Try figuring out what module is breaking it. `python -c 'import _ssl'; python -c 'import struct'; python -c 'import ctypes'` are some common build failure symptoms (ctypes you already had issues with).
* If that fails, try dtruss: `sudo dtruss /path/to/pyenv/python -E -m ensurepip` and post output.

Basically, we're trying to find what it wants to link to that it isn't finding.

**Comment by dave2wave** [comment]
@merlimat - Any reason not to Squash and Merge? (once the CI passes?)

@eolivelli - i've noticed that @zbentley is active with on Pulsar in the Homebrew project.

**Comment by zbentley** [comment]
@dave2wave, @merlimat and I paired on this awhile ago and had isolated build-repro issues down to something related to Python locale preventing Pyenv pythons from building on Matteo's setup. I can try a few "see what sticks" standard-issue fixes like LC_ALL exports etc. to see if that unblocks him and/or others. 

To do more than that, I need more info about specific build failures on builders' machines in order to give confidence in these changes; I'm happy to collaborate here or synchronously via Zoom; feel free to reach out to zac at klaviyo.com if interested.

Others on this thread: have you had luck (good or bad) building with these changes?

**Comment by zbentley** [comment]
@merlimat `pyenv` proved not to be worth it (and was responsible for the locale issues we encountered). This PR and its description have been updated to use rosetta but not pyenv to improve the build reproducibility. PTAL.

**Comment by tisonkun** [comment]
Closed. This looks good but the code has been moved to https://github.com/apache/pulsar-client-python.
